src/System.Threading.Thread/src/System/Threading/Thread.cs

```diff
@@ -190,14 +191,44 @@ public static IPrincipal CurrentPrincipal
         {
             get
             {
-                return CurrentThread._principal ?? (CurrentThread._principal = AppDomain.CurrentDomain.GetThreadPrincipal());
+                if (CurrentThread._principal is null)
+                {
+                    IPrincipal principal = AppDomain.CurrentDomain.GetThreadPrincipal();
+                    if (principal is null)
+                    {
+                        return null;
+                    }
+                    else
+                    {
+                        SetAsyncLocalPrincipal(principal);
+                    }
+                }
+                return CurrentThread._principal;
             }
             set
             {
-                CurrentThread._principal = value;
+                SetAsyncLocalPrincipal(value);
             }
         }
 
+        private static void SetAsyncLocalPrincipal(IPrincipal principal)
```

Comment:
This should not need to be separate method. The getter implementation above can do `Principal = principal;` to call this.